tests/test_date.py

```diff
@@ -118,163 +191,225 @@ def test_should_one_date_for_each_month(self, low, high, expected_results):
+            try:
+                result = date.get_intersecting_periods(self.period_low, self.period_high, period=period).next()
+            except Exception as error:
+                result = error
+            finally:
+                self.results.append(result)
+
+    def then_should_reject_easily_mistaken_dateutil_arguments(self, periods):
+        for p in izip(self.results, periods):
+            self.assertIsInstance(p[0], ValueError)
+            self.assertEqual('Invalid period: {}'.format(p[1]), p[0].message)
+
+
+class ParseDateWithFormats(BaseTestCase):
+    def setUp(self):
+        super(ParseDateWithFormats, self).setUp()
+        self.dt_data = NotImplemented
+
+    def test_date_with_not_matching_format_is_not_parsed(self):
+        self.given_invalid_date('yesterday')
+        self.then_shouldnt_parse_invalid_date()
 
     def test_should_parse_date(self):
-        result = date.parse_with_formats('25-03-14', ['%d-%m-%y'])
-        self.assertEquals(datetime(2014, 3, 25).date(), result['date_obj'].date())
+        self.given_a_date(datetime(2014, 3, 25))
+        self.when_date_is_parsed_with_format('25-03-14', '%d-%m-%y')
+        self.then_parsed_date_is()
 
     def test_should_use_current_year_for_dates_without_year(self):
-        today = datetime.today()
+        self.given_current_year()
+        self.given_day(16)
+        self.given_month(9)
+        self.when_date_is_parsed_with_format('09.16', "%m.%d")
+        self.then_should_use_current_year_for_dates_without_year()
+
+    def test_should_use_date_for_dates_without_day(self):
+        self.given_a_date(datetime(2014, 8, 31))
+        self.given_now(datetime(2014, 8, 12))
+        self.given_utc_date()
+        self.when_date_is_parsed_with_format('August 2014', '%B %Y')
+        self.then_dates_is_not_none()
+        self.then_period_equal('month')
+        self.then_dates_are_equal()
+
+    def given_day(self, day):
+        self.day = day
+
+    def given_month(self, month):
+        self.month = month
+
+    def given_invalid_date(self, date_str):
+        self.invalid_date = date_str
+
+    def given_current_year(self):
+        self.year = datetime.today().year
 
-        result = date.parse_with_formats('09.16', ["%m.%d"])
-        self.assertEquals(datetime(today.year, 9, 16).date(), result['date_obj'].date())
+    def given_a_date(self, date_obj):
+        self.date_obj = date_obj.date()
 
-    def test_should_use_current_date_for_dates_without_day(self):
-        twelfth = datetime(2014, 8, 12)
-        datetime_mock = Mock(wraps=datetime)
-        datetime_mock.utcnow = Mock(return_value=twelfth)
+    def given_now(self, given_date):
+        self.datetime_mock = Mock(wraps=datetime)
+        self.datetime_mock.utcnow = Mock(return_value=given_date)
 
-        with patch('dateparser.date_parser.datetime', new=datetime_mock):
-            dt_data = date.parse_with_formats('August 2014', ['%B %Y'])
+    def given_utc_date(self):
```

Comment:
What is the point of having separate method for adding patch instead of patching inside `given_now`?